### ISSUE TITLE
Explicit instantiation of vector<FilterMatch>

### ISSUE THREAD

**Issue by hanzhuoran** [openmc-dev/openmc#1122]
In filter.h, an explicit instantiation of vector<FilterMatch> is added. 
However, when making on MacOS, there would be an error reporting duplicate symbols for `std::vector<openmc::FilterMatch>`.
If this instantiation is commented out, OpenMC can be built successfully. However, there would be a seg fault at the start of active cycles.

**Comment by paulromano** [comment]
What version of macOS are you running? And what compiler are you using?

**Comment by hanzhuoran** [comment]
macOS is 10.13.6.
I tried GNU version 5.5.0, 6.3.0, 7.2.0, 8.30 installed from Homebrew for x86_64-apple-darwin17.5.0.

**Comment by paulromano** [comment]
I was able to reproduce the problem and have a fix for it that I'll push to #1120.

**Comment by hanzhuoran** [comment]
At the beginning of active cycles, there is still a segmentation fault:11.

**Comment by paulromano** [comment]
I was able to reproduce this as well. Running it through lldb gives:
```
* thread #1, queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=1, address=0x0)
    frame #0: 0x00000001001c4bf0 libopenmc.dylib`::filter_match_bins_clear(FilterMatch *) [inlined] std::vector<int, std::allocator<int> >::clear(this=0x0000000000000000) at stl_vector.h:1386
   1383	       */
   1384	      void
   1385	      clear() _GLIBCXX_NOEXCEPT
-> 1386	      { _M_erase_at_end(this->_M_impl._M_start); }
   1387	
   1388	    protected:
   1389	      /**
```

So far, I haven't been able to come up with a solution or even a good idea of why this is happening. The only way of getting around the segfault right now is to 1) compile with `-Ddebug=on` or 2) disable OpenMP.

**Comment by paulromano** [comment]
So after a little more digging, I have a legitimate workaround (that I don't really understand). I found that the problem is that `filter_match_pointer` was returning invalid pointers. To be more precise, the address of `filter_matches[0]` was showing up as `0x0` (and the size of `filter_matches` shows up as zero even after it has been resized). Strangely, moving the definition of `filter_match_pointer` from filter.cpp to simulation.cpp (where `filter_matches` is resized) seems to fix this. Really have no idea what's going on -- @smharper any thoughts?

**Comment by hanzhuoran** [comment]
I can work around by either compiling with -Ddebug=on or disabling OpenMP.

**Comment by smharper** [comment]
Hmm, I'm pretty stumped, too.  Although I have had troubles before with `extern "C"` blocks.  Does it work in filter.cpp if you define it as
```c++
extern "C" FilterMatch* filter_match_pointer(int indx) {...}
```
rather than
```c++
extern "C" {
  FilterMatch* filter_match_pointer(int indx) {...}
}
```

**Comment by paulromano** [comment]
@smharper nope, unfortunately that doesn't help. I thought maybe there was some weirdness with the symbols in simulation.cpp and filter.cpp, but running `nm` seems to indicate that things are fine:
```sh
~/openmc/build $ nm CMakeFiles/libopenmc.dir/src/simulation.cpp.o | grep filter_matches
                 U ___emutls_v._ZN6openmc10simulation14filter_matchesE
~/openmc/build $ nm CMakeFiles/libopenmc.dir/src/tallies/filter.cpp.o | grep filter_matches
0000000000008700 D ___emutls_v._ZN6openmc10simulation14filter_matchesE
```
This is pretty much what you'd expect -- same symbol appears twice but only defined in one translation unit.

The fact that this doesn't happen for debug builds is making me lean towards just declaring this a compiler bug... Anyway, I did put the odd workaround (moving `filter_match_pointer` to simulation.cpp) in my open PR.

**Comment by paulromano** [comment]
This should be fixed now with the merging of #1120. @hanzhuoran feel free to reopen if you are still experiencing problems.
